overlord/snapstate/handlers.go

```diff
@@ -1244,14 +1276,22 @@ func (m *SnapManager) undoCopySnapData(t *state.Task, _ *tomb.Tomb) error {
 	return nil
 }
 
-// writeMigrationStatus writes the state and sequence file (if they exist).
-// This must be called after the migration undo procedure is done since only
-// then do we know the actual final state of the migration.
-func writeMigrationStatus(t *state.Task, snapst *SnapState, snapName string) error {
+// writeMigrationStatus writes the SnapSetup, state and sequence file (if they
+// exist). This must be called after the migration undo procedure is done since
+// only then do we know the actual final state of the migration.
+func writeMigrationStatus(t *state.Task, snapst *SnapState, snapsup *SnapSetup) error {
 	st := t.State()
+	snapName := snapsup.InstanceName()
+
+	st.Lock()
+	err := SetTaskSnapSetup(t, snapsup)
+	st.Unlock()
```

Comment:
Can we just `defer st.Unlock` above? There are no expensive backend ops here afaict? Or is it problematic due to `writeSeqFile`?

Reply:
Yes, it was because of the file write but the impact might be negligible. Looking at handlers.go, I see we do lots of file operations while holding the lock so it's probably fine to defer here as well. Thanks!